### ISSUE TITLE
g++ builds failing after recent change

### ISSUE THREAD

**Issue by jkeenan** [Perl/perl5#22032]
Yesterday, I synched CPAN distribution Compress-Raw-Zlib-2.208 into blead (see fd0a86c119bad377b2d6b9ee4ecfc27c8d34d126).  It tested successfully on Ubuntu Linux, where `gcc` in the C-compiler, and on FreeBSD, where `clang` is the default C-compiler.  However, overnight blead began to [fail to build](http://perl.develop-help.com/?b=blead) on smoke-testing rigs that are using `g++` or `clang++` as the compiler.

Today I built with `g++` on both OSes.
```
sh ./Configure -des -Dusedevel -Dcc=g++ && make test_prep
```
Both builds failed during `make`.  On Linux (`g++-11), the first error seen in  the `make test_prep` output was this:
```
deflate.c: In function ‘int Perl_crz_deflateCopy(z_streamp, z_streamp)’:
deflate.c:1298:13: error: invalid conversion from ‘void*’ to ‘Perl_crz_Bytef*’ {aka ‘unsigned char*’} [-fpermissive]
 1298 |     zmemcpy((voidpf)dest, (voidpf)source, sizeof(z_stream));
      |             ^~~~~~~~~~~~
      |             |
      |             void*
In file included from ./zlib-src/deflate.h:16,
                 from deflate.c:52:
./zlib-src/zutil.h:214:38: note:   initializing argument 1 of ‘void zmemcpy(Perl_crz_Bytef*, const Perl_crz_Bytef*, Perl_crz_uInt)’
```
This error appeared repeatedly before `make test_prep` gave up:
```
$ zgrep -n 'error:' e8fa1e7b9f.linux.unthreaded.g++-11.maketp.output.txt.gz 
698:deflate.c:1298:13: error: invalid conversion from ‘void*’ to ‘Perl_crz_Bytef*’ {aka ‘unsigned char*’} [-fpermissive]
707:deflate.c:1298:27: error: invalid conversion from ‘void*’ to ‘const Perl_crz_Bytef*’ {aka ‘const unsigned char*’} [-fpermissive]
717:deflate.c:1303:13: error: invalid conversion from ‘void*’ to ‘Perl_crz_Bytef*’ {aka ‘unsigned char*’} [-fpermissive]
726:deflate.c:1303:25: error: invalid conversion from ‘void*’ to ‘const Perl_crz_Bytef*’ {aka ‘const unsigned char*’} [-fpermissive]
736:deflate.c:1318:13: error: invalid conversion from ‘void*’ to ‘Perl_crz_Bytef*’ {aka ‘unsigned char*’} [-fpermissive]
745:deflate.c:1318:31: error: invalid conversion from ‘void*’ to ‘const Perl_crz_Bytef*’ {aka ‘const unsigned char*’} [-fpermissive]
755:deflate.c:1319:13: error: invalid conversion from ‘void*’ to ‘Perl_crz_Bytef*’ {aka ‘unsigned char*’} [-fpermissive]
764:deflate.c:1319:31: error: invalid conversion from ‘void*’ to ‘const Perl_crz_Bytef*’ {aka ‘const unsigned char*’} [-fpermissive]
780:deflate.c:1298:13: error: invalid conversion from ‘void*’ to ‘Perl_crz_Bytef*’ {aka ‘unsigned char*’} [-fpermissive]
789:deflate.c:1298:27: error: invalid conversion from ‘void*’ to ‘const Perl_crz_Bytef*’ {aka ‘const unsigned char*’} [-fpermissive]
799:deflate.c:1303:13: error: invalid conversion from ‘void*’ to ‘Perl_crz_Bytef*’ {aka ‘unsigned char*’} [-fpermissive]
808:deflate.c:1303:25: error: invalid conversion from ‘void*’ to ‘const Perl_crz_Bytef*’ {aka ‘const unsigned char*’} [-fpermissive]
818:deflate.c:1318:13: error: invalid conversion from ‘void*’ to ‘Perl_crz_Bytef*’ {aka ‘unsigned char*’} [-fpermissive]
827:deflate.c:1318:31: error: invalid conversion from ‘void*’ to ‘const Perl_crz_Bytef*’ {aka ‘const unsigned char*’} [-fpermissive]
837:deflate.c:1319:13: error: invalid conversion from ‘void*’ to ‘Perl_crz_Bytef*’ {aka ‘unsigned char*’} [-fpermissive]
846:deflate.c:1319:31: error: invalid conversion from ‘void*’ to ‘const Perl_crz_Bytef*’ {aka ‘const unsigned char*’} [-fpermissive]
```
Similar output on FreeBSD-13 (`g++12).

@pmqs, can you take a look?  Thanks.

**Comment by pmqs** [comment]
Thanks. Already reported via https://github.com/pmqs/Compress-Raw-Zlib/issues/31

I'll get it sorted.

**Comment by pmqs** [comment]
Fixed in version 2.09. Just uploaded to CPAN. Also on GitHub at https://github.com/pmqs/Compress-Raw-Zlib/releases/tag/v2.209

I *always* forget to test the C++ use-case for this module. 

Previously I had suggested a C++ build for the GitHub Perl Action (see https://github.com/shogo82148/actions-setup-perl/issues/1294)  to allow me to test this use-case but that didn't go anywhere.

In the meantime, I've just added  `-Wc++-compat -Werror` to GitHub matrix I'm using on Linux. Means I shouldn't miss it next time zlib does an update.

**Comment by jkeenan** [comment]
Should be fixed in blead by:
```
commit b07cb93233bb4392c5366e5dcbd0f7b22e3b0c3b (HEAD -> blead, origin/blead, origin/HEAD, jkeenan/tmpblead)
Author:     Paul Marquess <pmqs@cpan.org>
AuthorDate: Mon Feb 26 13:49:24 2024 -0500
Commit:     James E Keenan <jkeenan@cpan.org>
CommitDate: Mon Feb 26 13:50:16 2024 -0500

    cpan/Compress-Raw-Zlib - Update to version 2.209
    
      2.209 26 February 2024
    
          * version 2.209
            Mon Feb 26 15:54:39 2024 +0000
            b8c062949b7e89a2f1643825f6b66f79b2ba3b68
    
          * #31 Fix build failure with "c++-compat" warnings
            Mon Feb 26 15:20:07 2024 +0000
            d9af6f896accf755fa1ef486cf128fdf66314213
```
